src/typecheck.rs

```diff
@@ -529,62 +566,87 @@ fn type_check_(
             let src = TypeWrapper::Ptr(new_var(state.table));
             // TODO what to do here, this makes more sense to me, but it means let x = foo in bar
             // behaves quite different to (\x.bar) foo, worth considering if it's ok to type these two differently
-            // let src = TypeWrapper::The(AbsType::Dyn());
             let trg = TypeWrapper::Ptr(new_var(state.table));
             let arr = mk_tyw_arrow!(src.clone(), trg.clone());
 
-            unify(state, strict, ty, arr).map_err(|err| err.into_typecheck_err(state, rt.pos))?;
+            check_sub(state, strict, arr, ty).map_err(|err| err.into_typecheck_err(state, rt.pos))?;
 
             envs.insert(x.clone(), src);
-            type_check_(state, envs, strict, t, trg)
+            check_(state, envs, strict, t, trg)
         }
         Term::List(terms) => {
-            let ty_elts = TypeWrapper::Ptr(new_var(state.table));
-
-            unify(state, strict, ty, mk_typewrapper::list(ty_elts.clone()))
-                .map_err(|err| err.into_typecheck_err(state, rt.pos))?;
+            let ty_elts = match ty {
+                // The following allows to type heterogenuous lists with `List Dyn`.
+                //
+                // Consider the following example:
+                // `[1, "a"] : List Dyn`
+                //
+                // The generic code below in the second match branch would spin up a type `List _a` for
+                // a fresh `_a` and apply a subtyping constraint `List _a <: List Dyn`. But because
+                // `_a <: Dyn` doesn't constrain `_a`, the unification variable would stay unresolved,
+                // and the list would be checked against `List _a`. This would fail because of the
+                // incompatible constraints `_a = Num` and `_a = Str` following from typing `1 : _a` and
+                // `"a": _a`.
```

Comment:
As stated in my other comment, I think this stems from the fact that the infer + constrain is less general than actual checking mode, and this is what happens here in part.

However there is a more general issue lying underneath which is that inference never tries to determine upper/lower bounds, which makes checking more "powerful" (or rather, more accepting) than inference. Thus, adding an unannotated variable as a proxy will always  cause mismatches, and this is not limited to lists:

```
// All of the following in a typed block

let fst : forall a. a -> a -> a = fun x y => x in
fst 1 "a" // Err

(fst 1 "a" : Dyn) // Works: forces the instantiation of a to Dyn
let x = fst 1 "a" in (x : Dyn) // Err

// About this one I'm not sure, given how we encode if-then-else, but it should be that way, if it is consistent
if true then 0 else "a" // Err
(if true then 0 else "a" : Dyn) // Works (probably?)
let x = if true then 0 else "a" in (x : Dyn) // Err
```

Writing this, I wonder if this possible to exhibit the same kind of examples in, say, GHC: it seems the ingredients are a non-trivial subtyping relation, a bidirectional type checking, and an inference that is "weak" in some sense (compared to checking and with respect to subtyping).

Note that beside subtyping (and this PR), the same kind of behavior is already happening with polymorphism and generalization:

```
// in a typed block
(fun x => x) : forall a. a -> a // Ok
let f = fun x => x in (f : forall a. a -> a) // Err
```

IMHO, `let x = [1, "a"] in ...` and alike should never typecheck, whatever the context is. Because `Dyn` works in so many situations, if we start to infer it where type mismatches occur, we'll probably end up propagating `Dyn` everywhere and accepting dubious code.

In the end, I don't find this behavior too unreasonable. Maybe the correct CSEd version of your initial example should be `let x : List Dyn = [1, "a"]  in x`, which indeed typechecks, but introducing an intermediate variable without annotation do force this variable to be inferred, which changes the behavior of tyepchecking.